### PR TITLE
Architecture review P1/P2 hardening: tx guard, dead-letter cap, crawl budget, route tripwire

### PR DESCRIPTION
**Pull request by jmagar**
Implements the six P1/P2 correctness-hardening items from the architecture review, plus a fix for a pre-existing test flake surfaced along the way. Splits cleanly into reviewable commits.

## P1 — correctness
- **RAII transaction guard** (`feat(jobs)`): a single `ImmediateTx` guard owns every `BEGIN IMMEDIATE` transaction (enqueue / claim / reclaim / cleanup), replacing four duplicated `commit_or_rollback`/`rollback_best_effort` implementations. `commit`/`rollback`/`finish` consume the guard; an unsettled drop (early `?` return or panic) relies on the pool's `after_release` ROLLBACK hook, so a connection can never return to the pool mid-transaction and starve the worker lanes. The pool-starvation acquire-timing canary is folded into the guard so it covers every transactional path, not just claim.
- **Dead-letter cap** (`feat(jobs)`): `AXON_MAX_JOB_ATTEMPTS` (default 5, `0` = unlimited). The watchdog marks a stale `running` job `failed` once it has been reclaimed that many times instead of re-queueing it forever — bounds a job that crashes/hangs on every attempt. Reclaim logic moved to `src/jobs/store/reclaim.rs` to keep `store.rs` under the monolith cap.
- **Schema-version doc fix** (`docs`): code is at `PAYLOAD_SCHEMA_VERSION = 8`; CLAUDE.md claimed `seed_url` bumped it to 7 as the "current value". `seed_url` was a v5 field and current is v8 — corrected.

## P2 — maintainability
- **Config field-add docs** (`docs`): the "update test-helper struct literals" gotchas were stale — those whole-`Config` literals no longer exist. Test helpers use `Config::test_default()` (spreads `..Default::default()`), so a new field only needs `Config::default()`. Verified empirically: adding `max_job_attempts` compiled the full test binary with zero helper edits. Updated `CLAUDE.md`, `src/cli/CLAUDE.md`, `src/core/CLAUDE.md`.
- **Route/OpenAPI tripwire** (`test(web)`): `openapi_document_matches_openapi_route_inventory` already locked inventory↔OpenAPI, and `all_v1_rest_routes_reject_missing_auth` probes inventory→router. Added the missing direction (router→inventory): a `.route("/v1/…")`/`.nest("/v1/…")` added to `routing.rs` without a matching `rest_route_inventory()` entry now fails the suite.
- **Crawl budget at claim** (`fix(crawl)`): anchor the wall-clock deadline at job claim (before validation) and bound the DNS-resolving `validate_crawl_job_url` by that deadline, so a slow/hung lookup counts against `crawl_job_timeout_secs` instead of parking the lane.

## Incidental
- `test(vector)`: serialized three `ensure_payload_indexes` tests that read the process-global `AXON_QDRANT_PAYLOAD_INDEX_PROFILE` env var (a pre-existing race the sibling `#[serial]` core-profile test sets up; surfaced by the new tests perturbing scheduling).

## Verification
- `cargo fmt --check`, `cargo clippy --all-targets`: clean.
- Full lib suite: **3307 passed, 0 failed**.
- Monolith policy: all changed files under the effective-line cap.
- Release gate: `cli` bumped 5.17.0 → 5.18.0; `check-release-versions --mode pr` + `check-version-sync` pass.

New tests: `ImmediateTx` drop-safety + commit/rollback, reclaim dead-letter, the route-inventory tripwire.

---
## Summary by cubic
Hardens the SQLite job runtime to prevent pool starvation and cap infinite job retries, fixes crawl timeout ordering by putting URL validation under the same deadline, and toughens the router→OpenAPI tripwire. Bumps release to 5.18.0.

- **New Features**
  - `AXON_MAX_JOB_ATTEMPTS` (default 5, `0` = unlimited): the watchdog now dead-letters a stale `running` job after that many reclaims instead of re-queuing it forever.

- **Bug Fixes**
  - Replaced manual `BEGIN IMMEDIATE` code with an RAII `ImmediateTx` guard across enqueue/claim/reclaim/cleanup; adds `#[must_use]`, logs slow acquires, and treats “no transaction is active” rollbacks as no-ops to avoid pool poisoning.
  - Crawl jobs start their wall-clock deadline at claim and now bound URL validation (DNS) by that same deadline so hung lookups don’t park a lane.
  - Route→inventory tripwire now detects multi-line `.route`/`.nest` registrations, includes a self-test floor, and fails the suite if a `/v1/...` route is missing from the inventory.
  - Registered `AXON_MAX_JOB_ATTEMPTS` in `docs/reference/env-matrix.toml` and `config.example.toml`, and wired env/TOML parsing to fall back to `Config::default()` so the default lives in one place.

<sup>Written for commit 445c9159c655dfc611cf3f55ec4d99b43c4907e0. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/jmagar/axon/pull/257?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Release Notes – Version 5.18.0

* **New Features**
  * Added `AXON_MAX_JOB_ATTEMPTS` configuration to control retry limits for stale running jobs; jobs exceeding the limit are now permanently marked as failed instead of being re-queued indefinitely.

* **Bug Fixes**
  * Fixed crawl timeout behavior to anchor deadlines at job claim time, ensuring validation and DNS resolution steps are properly bounded.

* **Chores**
  * Version bumped to 5.18.0; improved SQLite connection pooling for enhanced stability.

### CODE REVIEW

File: src/core/config/parse/build_config/config_literal.rs [hunk 1]
```diff
@@ -395,6 +395,10 @@ fn populate_misc(
         .or(inputs.toml.workers.crawl_job_timeout_secs)
         .unwrap_or(7200)
         .clamp(0, 86_400);
+    cfg.max_job_attempts = parse_i64_env("AXON_MAX_JOB_ATTEMPTS")
+        .or(inputs.toml.workers.max_job_attempts)
+        .unwrap_or(5)
+        .clamp(0, 1_000) as u32;
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/core/config/parse/build_config/config_literal.rs]: Fixed in 35b16998 — now `.map(|v| v.clamp(0,1000) as u32).unwrap_or(cfg.max_job_attempts)`, so the default lives only in `Config::default()` and the parser no longer duplicates the `5` literal.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/jobs/store/reclaim.rs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+            );
+        }
+        tracing::info!(
+            table,
+            reclaimed = n,
+            "watchdog: reclaimed stale running jobs"
+        );
+    }
+    Ok(jobs)
+}
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/jobs/store/reclaim.rs:256]: Fixed in 35b16998 — extracted `finalize_reclaimed_jobs()`; `reclaim_stale_running_jobs_for_table_jobs` is now back under the 80-line cap (monolith check passes with no warning).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,26 @@
 # Changelog
 
+## [5.18.0] - 2026-06-21
+
+### Added
```
[the remark this answers is not in the document]
Comment (jmagar) [reply CHANGELOG.md:5]: Skipping: every existing CHANGELOG entry uses this same Keep-a-Changelog heading spacing, and markdownlint isn't a CI gate in this repo — changing only the 5.18.0 entry would make it inconsistent with the rest of the file.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/jobs/store/reclaim.rs [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    // SAFETY: `kind.table_name()` returns a compile-time `&'static str` from
+    // a closed enum dispatch; no caller-controlled value reaches `format!`.
+    // Status literals come from a closed enum too.
+    let table = kind.table_name();
+    let threshold = now_ms() - stale_threshold_ms;
+    let reclaimed_at = now_ms();
+    let mut tx = ImmediateTx::begin(pool).await?;
+    let reclaimed_rows: Vec<ReclaimedRunningRow> = match sqlx::query_as(&format!(
+        "SELECT id, active_attempt_id, progress_json, result_json, attempt_count FROM {} WHERE status='running' AND updated_at < ?",
+        table
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/jobs/store/reclaim.rs:156]: Skipping: raw status literals in SQL strings are the established pattern across all job-table SQL (`store.rs`, `ops/lifecycle.rs`, `query.rs`). This code moved that pattern unchanged out of `store.rs` rather than introducing it; converting SQL-string state to `JobStatus::*.as_str()` is a codebase-wide change out of scope for this PR.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/jobs/workers/runners/crawl.rs [hunk 5]
```diff
@@ -211,6 +217,28 @@ async fn current_attempt_id(
...
+/// resolves DNS, so a hung lookup must not park the lane: when a deadline is set
+/// and elapses before validation returns, the crawl-timeout error is surfaced
+/// instead. Because `run_crawl_job` anchors the budget *before* calling this, a
+/// slow validation counts against `crawl_job_timeout_secs` rather than handing the
+/// engine a fresh full budget. A `None` deadline (timeout disabled) passes through.
+async fn validate_within_budget(
+    url: &str,
+    cancel_token: Option<&CancellationToken>,
+    budget: CrawlBudget,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/jobs/workers/runners/crawl.rs:230]: Skipping: `validate_within_budget` wraps `validate_crawl_job_url`, which returns `Box<dyn Error>` — matching it keeps the two consistent. Returning `CrawlGuardError` would force mapping SSRF/DNS validation failures into `CrawlGuardError::Engine`, mislabeling them as engine errors. The one typed value (Timeout) is boxed at the same boundary the wrapped fn's errors are.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
